test/boost/row_cache_test.cc

```diff
@@ -3951,6 +3952,148 @@ SEASTAR_TEST_CASE(test_scans_erase_dummies) {
+        s.delete_range(m1, query::full_clustering_range);
+        auto k1 = s.make_ckey(7);
+        s.add_row(m1, k1, "v1");
+
+        memtable_snapshot_source underlying(s.schema());
+        underlying.apply(m1);
+
+        cache_tracker tracker;
+        row_cache cache(s.schema(), snapshot_source([&] { return underlying(); }), tracker);
+
+        auto pr = dht::partition_range::make_singular(pkey);
+
+        {
+            auto slice = partition_slice_builder(*s.schema())
+                    .with_range(query::clustering_range::make_singular(k1))
+                    .build();
+            assert_that(cache.make_reader(s.schema(), semaphore.make_permit(), pr, slice))
+                    .produces(m1, slice.row_ranges(*s.schema(), pkey.key()));
+        }
+
+        auto misses_before = tracker.get_stats().row_misses;
+
+        {
+            auto slice = partition_slice_builder(*s.schema())
+                    .with_range(query::clustering_range::make_singular(k1))
+                    .build();
+            assert_that(cache.make_reader(s.schema(), semaphore.make_permit(), pr, slice))
+                    .produces(m1, slice.row_ranges(*s.schema(), pkey.key()));
+        }
+
+        BOOST_REQUIRE_EQUAL(misses_before, tracker.get_stats().row_misses);
+    });
+}
+
+SEASTAR_TEST_CASE(test_version_merging_with_range_tombstones_over_rowless_version) {
+    return seastar::async([] {
+        simple_schema s;
+        tests::reader_concurrency_semaphore_wrapper semaphore;
+
+        auto pkey = s.make_pkey("pk");
+        auto pr = dht::partition_range::make_singular(pkey);
+
+        memtable_snapshot_source underlying(s.schema());
+
+        mutation m1(s.schema(), pkey);
+        m1.partition().apply(s.new_tombstone());
+        underlying.apply(m1);
+
+        cache_tracker tracker;
+        row_cache cache(s.schema(), snapshot_source([&] { return underlying(); }), tracker);
+
+        // Populate cache
+        assert_that(cache.make_reader(s.schema(), semaphore.make_permit(), pr))
+                .produces(m1);
+
+        mutation m2(s.schema(), pkey);
+        s.delete_range(m2, s.make_ckey_range(7, 13));
+        s.delete_range(m2, s.make_ckey_range(9, 17));
+
+        {
+            auto rd1 = cache.make_reader(s.schema(), semaphore.make_permit(), pr);
+            auto close_rd1 = deferred_close(rd1);
+            rd1.set_max_buffer_size(1); // To hold the snapshot
+            rd1.fill_buffer().get();
+
+            apply(cache, underlying, m2);
+
+            evict_one_row(tracker); // hits last dummy in oldest version.
+
+            assert_that(cache.make_reader(s.schema(), semaphore.make_permit(), pr))
+                    .produces(m1 + m2);
+
+            evict_one_row(tracker); // hits entry in the latest version.
+        }
+
+        tracker.cleaner().drain().get();
+
+        assert_that(cache.make_reader(s.schema(), semaphore.make_permit(), pr))
+                .produces(m1 + m2);
+    });
+}
```

Comment:
I stared at the test for 20 minutes trying to understand what's it supposed to test, but I failed. The commit message says `Test violation of the "oldest version are evicted first" when evicting last dummy`, the test name says `test_version_merging_with_range_tombstones_over_rowless_version`, and neither rings a bell when I look at the code of the test. Could I get some comments here?

Reply:
Sure, how about this:

```
// Tests the following scenario:
//
// Initial state:
//
//   v2: ==== <7> [entry2] ==== <9> === <13> ==== <last dummy>
//   v1: ======================================== <last dummy> [entry1]
//
// After two eviction events which evict entry1 and entry2, we should end up with:
//
//   v2: ---------------------- <9> === <13> ==== <last dummy>
//   v1: ---------------------------------------- <last dummy>
//
// last dummy entries are treated in a special way in rows_entry::on_evicted(), and there
// was a bug which didn't clear the continuity on last dummy when it was selected for eviction.
// As a result, the view was this:
//
//   v2: ---------------------- <9> === <13> ==== <last dummy>
//   v1: ======================================== <last dummy>
//
// This would violate the "older versions are evicted first" rule, which implies
// that when entry2 is evicted in v2, the range in which entry2 falls into in all older versions
// must be discontinuous. This won't hold if we don't clear continuity on last dummy in v1.
// As a result, the range into which entry2 falls into from the perspective of v2 snapshot
// would appear as continuous and <7> would be missing from the read result, because
// continuity of a snapshot is a union of continuous ranges in all versions.
//
// Reproduces https://github.com/scylladb/scylladb/issues/12451
```

I opened https://github.com/scylladb/scylladb/issues/12451 for master, which is also affected. This was less likely to occur before mutation_partition_v2, which explains why concurrent_reads_and_eviction didn't catch this, because the 
incorrect state was only temporary. When version were merged, the situation was fixed. With v2, merging preserves the cursor's view, so the problem persists.